source/validate_datarules.cpp

```diff
@@ -68,9 +192,36 @@ namespace libspirv {
 // (Data Rules subsection of 2.16.1 Universal Validation Rules)
 spv_result_t DataRulesPass(ValidationState_t& _,
                            const spv_parsed_instruction_t* inst) {
-  if (auto error = ValidateNumVecComponents(_, inst)) return error;
-
-  // TODO(ehsan): add more data rules validation here.
+  switch (inst->opcode) {
+    case SpvOpTypeVector: {
+      if (auto error = ValidateVecNumComponents(_, inst)) return error;
+      break;
+    }
+    case SpvOpTypeFloat: {
+      if (auto error = ValidateFloatSize(_, inst)) return error;
+      break;
+    }
+    case SpvOpTypeInt: {
+      if (auto error = ValidateIntSize(_, inst)) return error;
+      break;
+    }
+    case SpvOpTypeMatrix: {
+      if (auto error = ValidateMatrixDataType(_, inst)) return error;
+      if (auto error = ValidateMatrixNumCols(_, inst)) return error;
+      break;
+    }
+    case SpvOpSpecConstant: {
```

Comment:
Generally, just add the TODO.   You never know how long it will take to get a review.  :-)